imap_processing/ultra/l1b/lookup_utils.py

```diff
@@ -281,19 +282,144 @@ def get_geometric_factor(
+    tuple
+        Scattering a and g values corresponding to the given theta and phi values.
+    """
+    filename = f"l1b-{instrument_id}sensor-scattering-calibration"
+    theta_grid = pd.read_csv(
+        ancillary_files[filename], header=None, skiprows=7, nrows=241
+    ).to_numpy(dtype=float)
+    phi_grid = pd.read_csv(
+        ancillary_files[filename], header=None, skiprows=249, nrows=241
+    ).to_numpy(dtype=float)
+    a_theta = pd.read_csv(
+        ancillary_files[filename], header=None, skiprows=491, nrows=241
+    ).to_numpy(dtype=float)
+    g_theta = pd.read_csv(
+        ancillary_files[filename], header=None, skiprows=733, nrows=241
+    ).to_numpy(dtype=float)
+    a_phi = pd.read_csv(
+        ancillary_files[filename], header=None, skiprows=975, nrows=241
+    ).to_numpy(dtype=float)
+    g_phi = pd.read_csv(
+        ancillary_files[filename], header=None, skiprows=1217, nrows=241
+    ).to_numpy(dtype=float)
+
+    # Assume uniform grids: extract 1D arrays from first row/col
+    theta_vals = theta_grid[0, :]  # columns represent theta
+    phi_vals = phi_grid[:, 0]  # rows represent phi
+
+    # Find nearest index in table for each input value
+    phi_idx = np.abs(phi_vals[:, None] - phi).argmin(axis=0)
+    theta_idx = np.abs(theta_vals[:, None] - theta).argmin(axis=0)
+
+    # Fetch a and g values at nearest (phi, theta) pairs
+    a_theta_val = a_theta[phi_idx, theta_idx]
+    g_theta_val = g_theta[phi_idx, theta_idx]
+    a_phi_val = a_phi[phi_idx, theta_idx]
+    g_phi_val = g_phi[phi_idx, theta_idx]
+
+    return np.column_stack([a_theta_val, g_theta_val]), np.column_stack(
+        [a_phi_val, g_phi_val]
+    )
+
+
+def pixels_below_fwhm_scattering_threshold(
+    theta_coeffs: np.ndarray,
+    phi_coeffs: np.ndarray,
+    energy: int,
+) -> np.ndarray:
+    """
+    Determine pixels below the FWHM scattering threshold.
+
+    For each phi and theta, calculate the FWHM using the formula:
+    FWHM = A*E^g
+    If Phi FWHM or Theta FWHM > the scattering requirements from the table above,
+    mask the instrument frame pixel.
+
+    Parameters
+    ----------
+    theta_coeffs : NDArray
+        Coefficients for theta FWHM calculation (a and g) for each pixel.
+    phi_coeffs : NDArray
+        Coefficients for phi FWHM calculation (a and g) for each pixel.
+    energy : int
+        Energy in keV.
+
+    Returns
+    -------
+    numpy.ndarray
+        Boolean array indicating pixels below the scattering threshold.
+    """
+    scattering_thresholds = UltraConstants.ULTRA_FWHM_SCATTERING_CULLING_THRESHOLDS
+    # Calculate FWHM for theta and phi
+    fwhm_theta = theta_coeffs[:, 0] * energy ** theta_coeffs[:, 1]
+    fwhm_phi = phi_coeffs[:, 0] * energy ** phi_coeffs[:, 1]
+
+    # Get the scattering threshold based on the energy
+    threshold = next(
+        threshold
+        for energy_range, threshold in scattering_thresholds.items()
+        if energy_range[0] <= energy < energy_range[1]
+    )
```

Comment:
OK, I see where this is used now and it is only a single lookup based on an integer energy and not an array-based lookup. You can probably ignore my other comment unless you like that approach better, but this seems fine to me as-is.